### PR TITLE
feat(arm64): enable single-binary builds

### PR DESCRIPTION
**Pull request by mudler**
**Description**

This PR tries to add a job to build on aarch64 by cross-compiling.

In the current state, it only builds the llama.cpp backend (non-cuda).

I've  tried compiling other backends and cuda but didn't had success yet.I'm leaving the CI configured for installing CUDA on the host, however that needs another iteration to figure out what's still wrong with it.

As a note, I've also added in the startup logics to allow to ship libraries as part of the backend assets and load those during start, under the `lib` directory, for instance, sibiling of the current directory we use to extract the grpc backends. 

In my tests with the binary it needed other libraries as well (which aren't part of this PR) in order to start llama.cpp correctly. This is not addressed in this PR, but I will likely follow up to leverage the lib loading to ship libraries along with the binary as well.

### CODE REVIEW

**Comment by sozercan** [comment]
@mudler looks like this needs cuda packages to be installed

another option i was thinking is that merge release into container builds (using qemu for arm64 builds), and have an optional container target that just copies the binaries into `scratch`, then output them with `--output`. however, i think the container builds take really long so this will take quite some time unless we this can be cached.

**Comment by mudler** [comment]
> @mudler looks like this needs cuda packages to be installed
> 
> another option i was thinking is that merge release into container builds (using qemu for arm64 builds), and have an optional container target that just copies the binaries into `scratch`, then output them with `--output`. however, i think the container builds take really long so this will take quite some time unless this can be cached.

yep that was actually what I had in mind as well.. but thought was worth giving a shot in any case as with cross-arch should supposedly be faster :crossed_fingers: 

however, as we build already an arm64 container maybe at the end that's what we have to end up doing to squeeze CI cycles

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -15,6 +15,91 @@ concurrency:
...
+          sudo apt-get install build-essential ffmpeg protobuf-compiler ccache
+          sudo apt-get install -qy binutils-aarch64-linux-gnu gcc-aarch64-linux-gnu g++-aarch64-linux-gnu
+      - name: Install CUDA Dependencies
+        run: |
+          curl -O https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/cross-linux-aarch64/cuda-keyring_1.1-1_all.deb
+          sudo dpkg -i cuda-keyring_1.1-1_all.deb
+          sudo apt-get update
+          sudo apt-get install -y cuda-cross-aarch64
+        env:
+          CUDA_VERSION: 12-3
```
Comment (sozercan) [inline .github/workflows/release.yaml:43]: nvidia isn't publishing cuda libraries for 12.3 for some reason, this need to be `12-4` or `12-5`

Comment (mudler) [reply .github/workflows/release.yaml]: we can start to think also to bundle that as part of the binary and then load libraries from there in runtime, that way we skip this annoying bit for users who are going to use static binaries

**Comment by mudler** [comment]
something else is off - probably the cross arch compiler isn't even picked up properly

**Comment by mudler** [comment]
long time I didn't do this - bare with me a bit. @sozercan any input is helpful as CI cycles takes a while

**Comment by mudler** [comment]
ok let's start step by step and have first an arm64 binary cross-compiled, will see what to do next with cuda - seems it's not being picked up correctly

File: .github/workflows/release.yaml (hunk 1 above)
Comment (mudler) [reply .github/workflows/release.yaml]: some of that logic implemented in https://github.com/mudler/LocalAI/pull/2490/commits/44120364c3aa17d5d5627b12e41e2d0c7468a3ff - it should allow us to carry over libs. I've just tried the arm64 builds and "almost" works in my env. llama.cpp doesn't start as there is a libstdc++ incompatibility with the host

**Comment by mudler** [comment]
aarch64 binary is compiled successfully for the aarch64, however in my test failed:

```
/tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback 
/tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by /tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback)
/tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by /tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback)
/tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback)
/tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback: /lib/aarch64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback)
/tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback: /lib/aarch64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback)
/tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback: /lib/aarch64-linux-gnu/libstdc++.so.6: version `CXXABI_1.3.13' not found (required by /tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback)
```

```
file /tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback 
/tmp/localai/backend_data/backend-assets/grpc/llama-cpp-fallback: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=dcff3a71166550acb940d410870543505d17aa94, for GNU/Linux 3.7.0, with debug_info, not stripped
```

so we are almost there. I'm trying to get a baseline now and merge what we have, we can iterate from there to add more backends to it

**Comment by mudler** [comment]
I want to try next to embed the libs (like nvidia too) into the final binary. will take a while to test, but if that works then we don't have to worry much more also in that specific case. If that won't work I could take it a little bit further in trying by pivoting root (from the user).

Anyhow, leaving libs handling for a follow-up, polishing now.

**Pull request merged**